src/coreclr/vm/jithelpers.cpp

```diff
@@ -2368,26 +2352,34 @@ HCIMPLEND
 // Currently, counter is a pointer into the Tier0 method stack
 // frame so we have exclusive access.
 
-void JIT_Patchpoint(int* counter, int ilOffset)
+extern "C" void JIT_PatchpointWorkerWorkerWithPolicy(TransitionBlock * pTransitionBlock)
 {
     // BEGIN_PRESERVE_LAST_ERROR;
     DWORD dwLastError = ::GetLastError();
 
     // This method may not return normally
-    STATIC_CONTRACT_GC_NOTRIGGER;
+    STATIC_CONTRACT_NOTHROW;
+    STATIC_CONTRACT_GC_TRIGGERS;
     STATIC_CONTRACT_MODE_COOPERATIVE;
 
+    PTR_PCODE pReturnAddress = (PTR_PCODE)(((BYTE*)pTransitionBlock) + TransitionBlock::GetOffsetOfReturnAddress());
+    PCODE ip = *pReturnAddress;
+    int* counter = *(int**)GetFirstArgumentRegisterValuePtr(pTransitionBlock);
+    int ilOffset = *(int*)GetSecondArgumentRegisterValuePtr(pTransitionBlock);
+    int hitCount = 1; // This will stay at 1 for forced transition scenarios, but will be updated to the actual hit count for normal patch points
+
     // Patchpoint identity is the helper return address
-    PCODE ip = (PCODE)_ReturnAddress();
 
     // Fetch or setup patchpoint info for this patchpoint.
     EECodeInfo codeInfo(ip);
     MethodDesc* pMD = codeInfo.GetMethodDesc();
     LoaderAllocator* allocator = pMD->GetLoaderAllocator();
     OnStackReplacementManager* manager = allocator->GetOnStackReplacementManager();
     PerPatchpointInfo * ppInfo = manager->GetPerPatchpointInfo(ip);
-    PCODE osrMethodCode = (PCODE)NULL;
+
     bool isNewMethod = false;
+    PCODE osrMethodCode = (PCODE)NULL;
+
 
     // In the current prototype, counter is shared by all patchpoints
```

Comment:
```suggestion
    // Counter is shared by all patchpoints
```
This is not a prototype anymore.